tests/generic_relations_regress/tests.py

```diff
@@ -332,3 +333,22 @@ def test_generic_reverse_relation_exclude_filter(self):
         self.assertSequenceEqual(qs, [link2])
         qs = Link.objects.exclude(places__name="Test Place 1")
         self.assertSequenceEqual(qs, [link2])
+
+    def test_ticket_34816_generic_fk_crash(self):
+        """
+        Ensure pk type change is handled when using GenericForeignKey.
+        """
+        board = Board.objects.create(name="some test")
+        oddrel = OddRelation1.objects.create(name="clink")
+        charlink = CharLink.objects.create(content_object=oddrel)
+        charlink = CharLink.objects.get(pk=charlink.pk)
+        self.assertEqual(type(charlink.content_object), OddRelation1)
+        self.assertEqual(
+            charlink.content_object, charlink._state.fields_cache["content_object"]
+        )
```

Comment:
Not sure if these two assertion against `charlink._state.fields_cache` are required as the issue manifests itself at a higher level abtraction.

Reply:
*I had added them to ensure that the `content_object` in cache is updated when accessed https://github.com/django/django/blob/254df3a3bbc2f4df51f9e2070ab2e214553f67d0/django/contrib/contenttypes/fields.py#L256*

Deleting them as deemed unnecessary.

Reply:
Deleted cache assertions.